src/lib.rs

```diff
@@ -314,40 +316,41 @@ impl GapfillDeltaTransition {
 // Note that for performance, this aggregate is parallel-izable, combinable, and does not expect ordered inputs.
 #[pg_extern(immutable, parallel_safe)]
 pub fn vector_selector_transition(
+    state: Internal,
+    start_time: TimestampTz,
+    end_time: TimestampTz,
+    bucket_width: Milliseconds,
+    lookback: Milliseconds,
+    time: TimestampTz,
+    value: f64,
     fcinfo: pg_sys::FunctionCallInfo,
-) -> Option<Internal<VectorSelector>> {
-    let fcinfo = unsafe { fcinfo.as_mut() }.unwrap();
-    let nargs = fcinfo.nargs;
-    let len = std::mem::size_of::<pg_sys::NullableDatum>() * nargs as usize;
-    let args = unsafe { fcinfo.args.as_slice(len) };
-
-    let state = unsafe { Internal::from_datum(args[0].value, args[0].isnull, pg_sys::INTERNALOID) };
-    let start_time =
-        unsafe { TimestampTz::from_datum(args[1].value, args[1].isnull, pg_sys::TIMESTAMPTZOID) }
-            .unwrap_or_else(|| error!("start_time is null"));
-    let end_time =
-        unsafe { TimestampTz::from_datum(args[2].value, args[2].isnull, pg_sys::TIMESTAMPTZOID) }
-            .unwrap_or_else(|| error!("end_time is null"));
-
-    let bucket_width =
-        unsafe { Milliseconds::from_datum(args[3].value, args[3].isnull, pg_sys::INT8OID) }
-            .unwrap_or_else(|| error!("bucket_width is null"));
-
-    let lookback =
-        unsafe { Milliseconds::from_datum(args[4].value, args[4].isnull, pg_sys::INT8OID) }
-            .unwrap_or_else(|| error!("lookback is null"));
-
-    let time =
-        unsafe { TimestampTz::from_datum(args[5].value, args[5].isnull, pg_sys::TIMESTAMPTZOID) }
-            .unwrap_or_else(|| error!("time is null"));
-
-    let value = unsafe { f64::from_datum(args[6].value, args[6].isnull, pg_sys::FLOAT8OID) }
-        .unwrap_or_else(|| error!("value is null"));
-
+) -> Internal {
+    vector_selector_transition_inner(
+        unsafe { state.to_inner() },
+        start_time.into(),
```

Comment:
will errors here be useful? The old code was explicit in what errors were thrown (e.g. error!("start_time is null"))

Reply:
This is a really good question, and I learned something by answering it (or maybe re-learned something I had already forgotten).

The main reason why I refactored this block with the `error!()` calls was because it was manually unpacking the args from fcinfo. When using the `#[pg_extern]` macro this is not necessary. In fact the macro re-writes the function and auto-generates _the same_ unpacking code which was implemented manually, _with_ the error warnings.

It's possible to expand macros and inspect the generated code. The following is the auto-generated code for the new function signature (there is a fair amount of magic going on here, but basically `pgx` autogenerates a `vector_selector_transition_wrapper` function which wraps `vector_selector_transition`):

```
pub unsafe extern "C" fn vector_selector_transition_wrapper(fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum {
    #[allow(non_snake_case)]
    #[allow(unused_variables)]
    #[allow(clippy::missing_safety_doc)]
    #[allow(clippy::redundant_closure)]
    unsafe fn vector_selector_transition_wrapper_inner(fcinfo: pg_sys::FunctionCallInfo) -> pg_sys::Datum {
        let result = {
            let state_ = pgx::pg_getarg::<Internal>(fcinfo, 0usize).unwrap_or_else(|| panic!("{} is null", stringify! {state }));
            let start_time_ = pgx::pg_getarg::<TimestampTz>(fcinfo, 1usize).unwrap_or_else(|| panic!("{} is null", stringify! {start_time }));
            let end_time_ = pgx::pg_getarg::<TimestampTz>(fcinfo, 2usize).unwrap_or_else(|| panic!("{} is null", stringify! {end_time }));
            let bucket_width_ = pgx::pg_getarg::<Milliseconds>(fcinfo, 3usize).unwrap_or_else(|| panic!("{} is null", stringify! {bucket_width }));
            let lookback_ = pgx::pg_getarg::<Milliseconds>(fcinfo, 4usize).unwrap_or_else(|| panic!("{} is null", stringify! {lookback }));
            let time_ = pgx::pg_getarg::<TimestampTz>(fcinfo, 5usize).unwrap_or_else(|| panic!("{} is null", stringify! {time }));
            let value_ = pgx::pg_getarg::<f64>(fcinfo, 6usize).unwrap_or_else(|| panic!("{} is null", stringify! {value }));
            let fcinfo_ = fcinfo;
            vector_selector_transition(state_, start_time_, end_time_, bucket_width_, lookback_, time_, value_, fcinfo )
        };
        result.into_datum().unwrap_or_else(|| panic!("returned Datum was NULL"))
    }
    pg_sys::guard::guard(|| vector_selector_transition_wrapper_inner(fcinfo ))
}

```

Reply:
Ah, I see that it doesn't `error!()`, but `panic!()`. Do you know if the unpacking with `errror!()` was done on purpose to avoid `panic!()`?

Reply:
I don't think that's purposeful panic should be fine. and the errors look ok. Can we just add a test for this.